File: go/idkollen/freja.go
package idkollen

import (
	"context"
	"encoding/json"
	"fmt"
	"time"
)

// --- Freja request types ---

type FrejaAuthRequest struct {
	Ssn                  *string `json:"ssn,omitempty"`
	CallbackUrl          *string `json:"callbackUrl,omitempty"`
	MinRegistrationLevel *string `json:"minRegistrationLevel,omitempty"`
	OrgNumber            *string `json:"orgNumber,omitempty"`
	RequestAddress       *bool   `json:"requestAddress,omitempty"`
	RefId                *string `json:"refId,omitempty"`
}

type FrejaBackchannelAuthRequest struct {
	Ssn                  string  `json:"ssn"`
	Country              string  `json:"country"`
	CallbackUrl          *string `json:"callbackUrl,omitempty"`
	MinRegistrationLevel *string `json:"minRegistrationLevel,omitempty"`
	OrgNumber            *string `json:"orgNumber,omitempty"`
	RequestAddress       *bool   `json:"requestAddress,omitempty"`
	RefId                *string `json:"refId,omitempty"`
}

type FrejaSignRequest struct {
	Text                 string  `json:"text"`
	Ssn                  *string `json:"ssn,omitempty"`
	CallbackUrl          *string `json:"callbackUrl,omitempty"`
	MinRegistrationLevel *string `json:"minRegistrationLevel,omitempty"`
	OrgNumber            *string `json:"orgNumber,omitempty"`
	RequestAddress       *bool   `json:"requestAddress,omitempty"`
	RefId                *string `json:"refId,omitempty"`
}

type FrejaBackchannelSignRequest struct {
	Ssn                  string  `json:"ssn"`
	Country              string  `json:"country"`
	Text                 string  `json:"text"`
	CallbackUrl          *string `json:"callbackUrl,omitempty"`
	MinRegistrationLevel *string `json:"minRegistrationLevel,omitempty"`
	OrgNumber            *string `json:"orgNumber,omitempty"`
	RequestAddress       *bool   `json:"requestAddress,omitempty"`
	RefId                *string `json:"refId,omitempty"`
}

// --- Freja status types ---

type FrejaStatus interface {
	isFrejaStatus()
}

type FrejaPending struct {
	Id             string  `json:"id"`
	RefId          *string `json:"refId"`
	AutoStartToken string  `json:"autoStartToken"`
	QrData         string  `json:"qrData"`
}

type FrejaCompleted struct {
	Id                   string  `json:"id"`
	RefId                *string `json:"refId"`
	Ssn                  string  `json:"ssn"`
	Country              string  `json:"country"`
	Name                 string  `json:"name"`
	GivenName            string  `json:"givenName"`
	Surname              string  `json:"surname"`
	Address              *string `json:"address"`
	CompanySignatoryText *string `json:"companySignatoryText"`
}

type FrejaFailed struct {
	Id    string  `json:"id"`
	RefId *string `json:"refId"`
	Error string  `json:"error"`
}

func (*FrejaPending) isFrejaStatus()   {}
func (*FrejaCompleted) isFrejaStatus() {}
func (*FrejaFailed) isFrejaStatus()    {}

type frejaStatusEnvelope struct {
	Status string `json:"status"`
}

func unmarshalFrejaStatus(data []byte) (FrejaStatus, error) {
	var env frejaStatusEnvelope
	if err := json.Unmarshal(data, &env); err != nil {
		return nil, err
	}
	switch env.Status {
	case "PENDING":
		var v FrejaPending
		if err := json.Unmarshal(data, &v); err != nil {
			return nil, err
		}
		return &v, nil
	case "COMPLETED":
		var v FrejaCompleted
		if err := json.Unmarshal(data, &v); err != nil {
			return nil, err
		}
		return &v, nil
	case "FAILED":
		var v FrejaFailed
		if err := json.Unmarshal(data, &v); err != nil {
			return nil, err
		}
		return &v, nil
	default:
		return nil, fmt.Errorf("unknown freja status: %s", env.Status)
	}
}

// --- Freja endpoint ---

type FrejaEndpoint struct {
	client *IdkollenClient
}

func (e *FrejaEndpoint) Auth(ctx context.Context, req FrejaAuthRequest) (FrejaStatus, error) {
	var raw json.RawMessage
	if err := e.client.post(ctx, "/v3/freja/auth", req, &raw); err != nil {
		return nil, err
	}
	return unmarshalFrejaStatus(raw)
}

func (e *FrejaEndpoint) BackchannelAuth(ctx context.Context, req FrejaBackchannelAuthRequest) (FrejaStatus, error) {
	var raw json.RawMessage
	if err := e.client.post(ctx, "/v3/freja/backchannel/auth", req, &raw); err != nil {
		return nil, err
	}
	return unmarshalFrejaStatus(raw)
}

func (e *FrejaEndpoint) Sign(ctx context.Context, req FrejaSignRequest) (FrejaStatus, error) {
	var raw json.RawMessage
	if err := e.client.post(ctx, "/v3/freja/sign", req, &raw); err != nil {
		return nil, err
	}
	return unmarshalFrejaStatus(raw)
}

func (e *FrejaEndpoint) BackchannelSign(ctx context.Context, req FrejaBackchannelSignRequest) (FrejaStatus, error) {
	var raw json.RawMessage
	if err := e.client.post(ctx, "/v3/freja/backchannel/sign", req, &raw); err != nil {
		return nil, err
	}
	return unmarshalFrejaStatus(raw)
}

func (e *FrejaEndpoint) AuthStatus(ctx context.Context, id string) (FrejaStatus, error) {
	var raw json.RawMessage
	if err := e.client.get(ctx, "/v3/freja/auth/"+id, &raw); err != nil {
		return nil, err
	}
	return unmarshalFrejaStatus(raw)
}

func (e *FrejaEndpoint) SignStatus(ctx context.Context, id string) (FrejaStatus, error) {
	var raw json.RawMessage
	if err := e.client.get(ctx, "/v3/freja/sign/"+id, &raw); err != nil {
		return nil, err
	}
	return unmarshalFrejaStatus(raw)
}

func (e *FrejaEndpoint) CancelAuth(ctx context.Context, id string) error {
	return e.client.delete(ctx, "/v3/freja/auth/"+id)
}

func (e *FrejaEndpoint) CancelSign(ctx context.Context, id string) error {
	return e.client.delete(ctx, "/v3/freja/sign/"+id)
}

func (e *FrejaEndpoint) WaitForAuth(ctx context.Context, id string, opts PollOptions) (FrejaStatus, error) {
	return pollFreja(ctx, opts, func() (FrejaStatus, error) { return e.AuthStatus(ctx, id) })
}

func (e *FrejaEndpoint) WaitForSign(ctx context.Context, id string, opts PollOptions) (FrejaStatus, error) {
	return pollFreja(ctx, opts, func() (FrejaStatus, error) { return e.SignStatus(ctx, id) })
}

func pollFreja(ctx context.Context, opts PollOptions, fn func() (FrejaStatus, error)) (FrejaStatus, error) {
	interval := opts.Interval
	if interval == 0 {
		interval = 2 * time.Second
	}
	for {
		status, err := fn()
		if err != nil {
			return nil, &WaitError{Err: err}
		}
		if _, ok := status.(*FrejaPending); !ok {
			return status, nil
		}
		select {
		case <-ctx.Done():
			return nil, &WaitError{Timeout: true}
		case <-time.After(interval):
		}
	}
}

File: ruby/lib/idkollen/freja.rb
# frozen_string_literal: true

module Idkollen
  module Freja
    class AuthRequest
      attr_reader :ssn, :callback_url, :min_registration_level, :org_number,
                  :request_address, :ref_id

      def initialize(ssn: nil, callback_url: nil, min_registration_level: nil,
                     org_number: nil, request_address: nil, ref_id: nil)
        @ssn = ssn; @callback_url = callback_url
        @min_registration_level = min_registration_level; @org_number = org_number
        @request_address = request_address; @ref_id = ref_id
      end

      def to_h
        {
          ssn: @ssn, callbackUrl: @callback_url,
          minRegistrationLevel: @min_registration_level, orgNumber: @org_number,
          requestAddress: @request_address, refId: @ref_id,
        }.compact
      end
    end

    class BackchannelAuthRequest
      attr_reader :ssn, :country, :callback_url, :min_registration_level,
                  :org_number, :request_address, :ref_id

      def initialize(ssn:, country:, callback_url: nil, min_registration_level: nil,
                     org_number: nil, request_address: nil, ref_id: nil)
        @ssn = ssn; @country = country; @callback_url = callback_url
        @min_registration_level = min_registration_level; @org_number = org_number
        @request_address = request_address; @ref_id = ref_id
      end

      def to_h
        {
          ssn: @ssn, country: @country, callbackUrl: @callback_url,
          minRegistrationLevel: @min_registration_level, orgNumber: @org_number,
          requestAddress: @request_address, refId: @ref_id,
        }.compact
      end
    end

    class SignRequest
      attr_reader :text, :ssn, :callback_url, :min_registration_level,
                  :org_number, :request_address, :ref_id

      def initialize(text:, ssn: nil, callback_url: nil, min_registration_level: nil,
                     org_number: nil, request_address: nil, ref_id: nil)
        @text = text; @ssn = ssn; @callback_url = callback_url
        @min_registration_level = min_registration_level; @org_number = org_number
        @request_address = request_address; @ref_id = ref_id
      end

      def to_h
        {
          text: @text, ssn: @ssn, callbackUrl: @callback_url,
          minRegistrationLevel: @min_registration_level, orgNumber: @org_number,
          requestAddress: @request_address, refId: @ref_id,
        }.compact
      end
    end

    class BackchannelSignRequest
      attr_reader :ssn, :country, :text, :callback_url, :min_registration_level,
                  :org_number, :request_address, :ref_id

      def initialize(ssn:, country:, text:, callback_url: nil, min_registration_level: nil,
                     org_number: nil, request_address: nil, ref_id: nil)
        @ssn = ssn; @country = country; @text = text; @callback_url = callback_url
        @min_registration_level = min_registration_level; @org_number = org_number
        @request_address = request_address; @ref_id = ref_id
      end

      def to_h
        {
          ssn: @ssn, country: @country, text: @text, callbackUrl: @callback_url,
          minRegistrationLevel: @min_registration_level, orgNumber: @org_number,
          requestAddress: @request_address, refId: @ref_id,
        }.compact
      end
    end

    class Pending
      attr_reader :id, :ref_id, :auto_start_token, :qr_data

      def initialize(id:, auto_start_token:, qr_data:, ref_id: nil)
        @id = id; @ref_id = ref_id
        @auto_start_token = auto_start_token; @qr_data = qr_data
      end

      def deconstruct_keys(_)
        { id: @id, ref_id: @ref_id, auto_start_token: @auto_start_token, qr_data: @qr_data }
      end

      def self.from_json(d)
        new(id: d['id'], ref_id: d['refId'],
            auto_start_token: d['autoStartToken'], qr_data: d['qrData'])
      end
    end

    class Completed
      attr_reader :id, :ref_id, :ssn, :country, :name, :given_name, :surname,
                  :address, :company_signatory_text

      def initialize(id:, ssn:, country:, name:, given_name:, surname:,
                     ref_id: nil, address: nil, company_signatory_text: nil)
        @id = id; @ref_id = ref_id; @ssn = ssn; @country = country
        @name = name; @given_name = given_name; @surname = surname
        @address = address; @company_signatory_text = company_signatory_text
      end

      def deconstruct_keys(_)
        { id: @id, ref_id: @ref_id, ssn: @ssn, country: @country,
          name: @name, given_name: @given_name, surname: @surname,
          address: @address, company_signatory_text: @company_signatory_text }
      end

      def self.from_json(d)
        new(
          id: d['id'], ref_id: d['refId'], ssn: d['ssn'], country: d['country'],
          name: d['name'], given_name: d['givenName'], surname: d['surname'],
          address: d['address'], company_signatory_text: d['companySignatoryText'],
        )
      end
    end

    class Failed
      attr_reader :id, :ref_id, :error

      def initialize(id:, error:, ref_id: nil)
        @id = id; @ref_id = ref_id; @error = error
      end

      def deconstruct_keys(_) = { id: @id, ref_id: @ref_id, error: @error }

      def self.from_json(d) = new(id: d['id'], ref_id: d['refId'], error: d['error'])
    end

    module Status
      def self.from_json(d)
        case d['status']
        when 'PENDING'   then Pending.from_json(d)
        when 'COMPLETED' then Completed.from_json(d)
        when 'FAILED'    then Failed.from_json(d)
        else raise ArgumentError, "Unknown freja status: #{d['status']}"
        end
      end
    end

    class Endpoint
      def initialize(transport)
        @transport = transport
      end

      def auth(req)                  = Status.from_json(@transport.post('/v3/freja/auth', req.to_h))
      def backchannel_auth(req)      = Status.from_json(@transport.post('/v3/freja/backchannel/auth', req.to_h))
      def sign(req)                  = Status.from_json(@transport.post('/v3/freja/sign', req.to_h))
      def backchannel_sign(req)      = Status.from_json(@transport.post('/v3/freja/backchannel/sign', req.to_h))
      def auth_status(id)            = Status.from_json(@transport.get("/v3/freja/auth/#{id}"))
      def sign_status(id)            = Status.from_json(@transport.get("/v3/freja/sign/#{id}"))
      def cancel_auth(id)            = @transport.delete("/v3/freja/auth/#{id}")
      def cancel_sign(id)            = @transport.delete("/v3/freja/sign/#{id}")

      def age_verification(req)
        Idkollen::AgeVerification::Status.from_json(@transport.post('/v3/freja/age-verification', req.to_h))
      end

      def age_verification_status(id)
        Idkollen::AgeVerification::Status.from_json(@transport.get("/v3/freja/age-verification/#{id}"))
      end

      def cancel_age_verification(id) = @transport.delete("/v3/freja/age-verification/#{id}")

      def wait_for_auth(id, opts: PollOptions.new) = poll(opts) { auth_status(id) }
      def wait_for_sign(id, opts: PollOptions.new) = poll(opts) { sign_status(id) }

      def wait_for_age_verification(id, opts: PollOptions.new)
        poll_age(opts) { age_verification_status(id) }
      end

      private

      def poll(opts)
        deadline = Time.now + opts.timeout
        loop do
          status = yield
          return status unless status.is_a?(Pending)
          raise WaitError.new(timeout: true) if Time.now >= deadline
          sleep opts.interval
        end
      end

      def poll_age(opts)
        deadline = Time.now + opts.timeout
        loop do
          status = yield
          return status unless status.is_a?(Idkollen::AgeVerification::Pending)
          raise WaitError.new(timeout: true) if Time.now >= deadline
          sleep opts.interval
        end
      end
    end
  end
end
